### PR TITLE
perf: hoist static vnodes

### PR DESCRIPTION
**Pull request by jodarove**
## Details

Note: this PR base branch is against #2758 for it to only include changes related to hoist static vnodes.

### Performance

With the performance test as they are, these are the results:

image

There's so much improvement in the `styled-component` tests, because the component has a static node `<div class="cmp-${i}">Hello world</div>`.

## Does this pull request introduce a breaking change?

* ✅ No, it does not introduce a breaking change.

## Does this pull request introduce an observable change?

* ✅ No, it does not introduce an observable change.

### CODE REVIEW

File: packages/@lwc/engine-core/src/framework/rendering.ts [hunk 1]
```diff
@@ -175,11 +176,41 @@ function mountElement(vnode: VElement, parent: ParentNode, anchor: Node | null)
...
     } = vnode;
 
+    if (vnode.isStatic) {
+        if (vnode.elm) {
+            // @todo: what to do about the owner for synthetic shadow and fallbackElmHook
+
+            // @todo: cloned subtree is not associated with the proper shadow. The restrictions and styles will be
+            //        ok.
+            //        maybe modify synthetic shadow and get the owner from ancestors?
+            vnode.elm = vnode.elm.cloneNode(true) as Element;
```
Comment (nolanlawson) [inline packages/@lwc/engine-core/src/framework/rendering.ts:186]: Is it necessary to always call `cloneNode(true)`? I think the very first time the VNode is rendered, the `elm` has never been used anywhere else, so it's safe to just use `vnode.elm`. This might actually be more performant, assuming that most static nodes are not used a 2nd time.

File: packages/@lwc/engine-core/src/framework/rendering.ts [hunk 2]
```diff
@@ -175,11 +176,41 @@ function mountElement(vnode: VElement, parent: ParentNode, anchor: Node | null)
...
+            insertNode(vnode.elm, parent, anchor);
+            return;
+        }
+
+        // @todo: hackalert: for now, let's mark all children as static and set the owner
+        const ch = vnode.children;
+        for (let i = 0, n = ch.length; i < n; i++) {
+            ch[i]!.isStatic = true;
+            ch[i]!.owner = owner;
+        }
```
Comment (nolanlawson) [inline packages/@lwc/engine-core/src/framework/rendering.ts:205]: Is this a hack because it could be done in the `tmpl` function?

File: packages/@lwc/engine-core/src/framework/rendering.ts [hunk 3]
```diff
@@ -712,6 +747,9 @@ function updateStaticChildren(c1: VNodes, c2: VNodes, parent: ParentNode) {
                 mount(n2, parent, anchor);
                 anchor = n2.elm!;
             }
+        } else if (n2 !== null) {
+            // it is hoisted
+            anchor = n2.elm!;
```
Comment (nolanlawson) [inline packages/@lwc/engine-core/src/framework/rendering.ts:752]: I don't understand this part, could you explain?

File: packages/@lwc/synthetic-shadow/src/faux-shadow/shadow-root.ts [hunk 4]
```diff
@@ -84,6 +85,20 @@ defineProperty(Node.prototype, KEY__SHADOW_RESOLVER, {
         (this as any)[KEY__SHADOW_RESOLVER_PRIVATE] = fn;
         // TODO [#1164]: temporary propagation of the key
         setNodeOwnerKey(this, (fn as any).nodeKey);
+
+        if (isTrue((this as any)[KEY__IS_STATIC_NODE])) {
+            const treeWalker = document.createTreeWalker(
+                this,
+                NodeFilter.SHOW_ELEMENT | NodeFilter.SHOW_COMMENT | NodeFilter.SHOW_TEXT
+            );
```
Comment (nolanlawson) [inline packages/@lwc/synthetic-shadow/src/faux-shadow/shadow-root.ts:93]: I'm pretty sure IE11 requires all four arguments to `createTreeWalker`. E.g. see:

https://github.com/salesforce/lwc/blob/e6b4c29a2140108cbc5b7dbf146b6a1d9c9ab392/packages/integration-karma/helpers/test-utils.js#L291-L296

File: packages/@lwc/template-compiler/src/codegen/codegen.ts [hunk 5]
```diff
@@ -51,16 +75,97 @@ const RENDER_APIS: { [primitive in RenderPrimitive]: RenderPrimitiveDefinition }
...
+            }
+            staticNodes.add(node);
+        }
+    }
+
+    root.children.forEach((childNode) => collectStaticNodes(childNode, root));
+
+    return staticNodes;
+}
+
```
Comment (nolanlawson) [inline packages/@lwc/template-compiler/src/codegen/codegen.ts:164]: Originally I implemented hoisting as a separate step in a separate file (`optimize.ts`) that runs after generation, but I think your approach of doing it directly during the initial generation is superior. It would be great to move `optimize.ts` into this phase as well.

File: packages/@lwc/template-compiler/src/codegen/codegen.ts [hunk 6]
```diff
@@ -172,11 +291,23 @@ export default class CodeGen {
...
+        const apiCall = this._renderApiCall(RENDER_APIS.text, [textConcatenation]);
+
+        return isHoisted ? this.genHoistedNode(apiCall) : apiCall;
+    }
+
+    genHoistedNode(expr: t.Expression): t.CallExpression {
+        this.hoistedNodes.push(expr);
+
+        return this._renderApiCall(RENDER_APIS.setOwner, [
+            t.identifier(`$hoisted${this.hoistedNodes.length}`),
```
Comment (nolanlawson) [inline packages/@lwc/template-compiler/src/codegen/codegen.ts:303]: In my hoisting, I used the prefix `stc_`. It would be nice to use consistent naming (although not a big deal).

**Review by nolanlawson — COMMENTED** [review]
This is really exciting! Awesome work!! 🎉

File: packages/@lwc/engine-core/src/framework/rendering.ts (hunk 1 above)
Comment (jodarove) [reply packages/@lwc/engine-core/src/framework/rendering.ts:186]: > Is it necessary to always call cloneNode(true)? 

yes, it may be that it is within a different component instances or within a foreach.

> assuming that most static nodes are not used a 2nd time.

yes it will, the hoisted vnode is shared across component instances, but component instances have different elms, that's why the cloneNode

File: packages/@lwc/engine-core/src/framework/rendering.ts (hunk 2 above)
Comment (jodarove) [reply packages/@lwc/engine-core/src/framework/rendering.ts:205]: well, it may not be a hack 😄. It can not be done in the template function because that will be executed in every rehydration, while here only executes just one time per component (!== component instance), when we need to create the element.

I marked it as a hack because I didn't give too much thought then, maybe there's a better way of achieving this.

The key idea is that we need to go through the regular mounting mechanism to mount the hoisting tree (just one time per module!) but only the root vnode is marked as static and has an owner, it may be the case that contains other vnodes, and we need to set the owner in them.

File: packages/@lwc/template-compiler/src/codegen/codegen.ts (hunk 5 above)
Comment (jodarove) [reply packages/@lwc/template-compiler/src/codegen/codegen.ts:164]: I need to think about it, though ill prefer to do it in a different PR.

from the top of my mind, I think there's a semantic difference between the 2 of them: `stc` is an optimization of the generated output, while `hoisted` is changing the generated output based on the input.

File: packages/@lwc/engine-core/src/framework/rendering.ts (hunk 1 above)
Comment (nolanlawson) [reply packages/@lwc/engine-core/src/framework/rendering.ts:186]: Let me rephrase my question: is it possible to avoid the `cloneNode` the _first_ time we reach this line of code? And then mark somewhere that the `elm` has already been used? E.g.

```js
if (vnode.elmWasUsed) {
  vnode.elm = vnode.elm.cloneNode(true)
} else {
  vnode.elmWasUsed = true
}
```

File: packages/@lwc/template-compiler/src/codegen/codegen.ts (hunk 6 above)
Comment (jodarove) [reply packages/@lwc/template-compiler/src/codegen/codegen.ts:303]: Well, this `hoisted` is different from `stc_`. It may be the case that you have both `hoisted$` and `stc_`. `hoisted` stores vnodes, while `stc_` store static expressions. I like that the names are different.

File: packages/@lwc/engine-core/src/framework/rendering.ts (hunk 1 above)
Comment (jodarove) [reply packages/@lwc/engine-core/src/framework/rendering.ts:186]: if the element was already used, then we would be doing a `patch`, and we do not do anything in patch (the first condition is `n1 !== n2 -> return`

File: packages/@lwc/engine-core/src/framework/rendering.ts (hunk 2 above)
Comment (nolanlawson) [reply packages/@lwc/engine-core/src/framework/rendering.ts:205]: Sorry, I didn't mean to say "in the `tmpl` function," I meant "can the compiled code do this?" Or can we do it in `setOwner`?

File: packages/@lwc/engine-core/src/framework/rendering.ts (hunk 3 above)
Comment (jodarove) [reply packages/@lwc/engine-core/src/framework/rendering.ts:752]: in line 736 there's a check `n2 !== n1`, which previous to this pr was only true when both nodes were `null`. now (remember, there's only 1 per module) that will be the also the case for hoisted vnodes, and we need to go via patch line 740, but there's nothing to patch, therefore we only need to update the anchor.

Comment (nolanlawson) [reply packages/@lwc/engine-core/src/framework/rendering.ts:752]: Could you add a code comment? I think this whole function is very complex and could benefit from it. 🙂

File: packages/@lwc/engine-core/src/framework/rendering.ts (hunk 2 above)
Comment (jodarove) [reply packages/@lwc/engine-core/src/framework/rendering.ts:205]: the compiled code can not do this, because there's no owner at the time that the vnode is created.

if we do it in `setOwner` we would be doing it for every re-hydration, vs just one time per module.

File: packages/@lwc/engine-core/src/framework/rendering.ts (hunk 1 above)
Comment (jodarove) [reply packages/@lwc/engine-core/src/framework/rendering.ts:186]: let me clarify, there's no way of knowing to which component instance this `vnode.elm` belongs to, we just know that the rendered element will be identical to the one that is previously set.

File: packages/@lwc/engine-core/src/framework/rendering.ts (hunk 2 above)
Comment (nolanlawson) [reply packages/@lwc/engine-core/src/framework/rendering.ts:201]: OK got it, thank you!

File: packages/@lwc/engine-core/src/framework/api.ts [hunk 7]
```diff
@@ -49,6 +49,14 @@ function addVNodeToChildLWC(vnode: VCustomElement) {
     ArrayPush.call(getVMBeingRendered()!.velements, vnode);
 }
 
+// [s]et [o]wner
+function so(vnode: VNode): VNode {
+    vnode.owner = getVMBeingRendered()!;
+    vnode.isStatic = true;
+
+    return vnode;
+}
```
[the remark this answers is not in the document]
Comment (jodarove) [reply packages/@lwc/engine-core/src/framework/api.ts:58]: I don't think we can run in cases where the owner is incorrectly set: this function (`so`) is invoked in the template evaluation, therefore, it will always have the correct owner. Notice that nothing is done for `patch`.

This PR does not need #2716, however, it can benefit from it, because then there's no need to have this function.

File: packages/integration-karma/test/api/sanitizeAttribute/x/xlink/xlink.html [hunk 8]
```diff
@@ -1,5 +1,5 @@
 <template>
     <svg>
-        <use xlink:href="/foo"></use>
+        <use xlink:href={xLink}></use>
```
[the remark this answers is not in the document]
Comment (jodarove) [reply packages/integration-karma/test/api/sanitizeAttribute/x/xlink/xlink.html:3]: this is because of what [the test ](https://github.com/salesforce/lwc/blob/master/packages/integration-karma/test/api/sanitizeAttribute/index.spec.js)does. the test reset `LWC.sanitizeAttribute`, but because the hoisted content is evaluated just once, then it never executes `sanitizeAttribute`.

File: packages/@lwc/template-compiler/src/codegen/codegen.ts [hunk 9]
```diff
@@ -51,16 +75,97 @@ const RENDER_APIS: { [primitive in RenderPrimitive]: RenderPrimitiveDefinition }
...
+            attributes,
+            directives,
+            properties,
+            listeners,
+        } = node;
+
+        // Note: the trick is that the static nodes elm can not be removed by themself, they will always be removed
+        //       as parent removal.
+
+        result &&= !isIf(parent); // the parent node is an if, therefore this it is the same as an inline if
```
[the remark this answers is not in the document]
Comment (jodarove) [reply packages/@lwc/template-compiler/src/codegen/codegen.ts:103]: Sure, the gist is that if this node is within a `template if:`, then it can't be static because it means that the diffing algo may remove the vnode.element directly, and at the point of removal, there's a 99% chance that the hoisted vnode does not contain the correct element to be removed.

File: packages/@lwc/engine-core/src/framework/api.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (caridy) [reply packages/@lwc/engine-core/src/framework/api.ts:58]: I'm not sure that's true... the owner might be used somewhere else... I think @pmdartus is correct here. This is what of those issues that I believe are fatal of trying to re-use an actual vnode.

[the remark this answers is not in the document]
Comment (jodarove) [reply packages/@lwc/engine-core/src/framework/api.ts]: which is that case? to the best of my knowledge, the vnodes are kept for reference only to do the diffing.

File: packages/@lwc/engine-core/src/framework/api.ts [hunk 10]
```diff
@@ -49,6 +49,14 @@ function addVNodeToChildLWC(vnode: VCustomElement) {
     ArrayPush.call(getVMBeingRendered()!.velements, vnode);
 }
 
+// [s]et [o]wner
+function so(vnode: VNode): VNode {
+    vnode.owner = getVMBeingRendered()!;
+    vnode.isStatic = true;
```
[the remark this answers is not in the document]
Comment (jodarove) [reply packages/@lwc/engine-core/src/framework/api.ts]: resolved in ed84e47a6137898b57b84fa98efe14b850c1ca69

File: packages/@lwc/engine-core/src/framework/vnodes.ts [hunk 11]
```diff
@@ -74,6 +75,7 @@ export interface VNodeData {
     readonly context?: Readonly<Record<string, Readonly<Record<string, any>>>>;
     readonly on?: Readonly<Record<string, (event: Event) => any>>;
     readonly svg?: boolean;
+    readonly isStatic?: boolean;
```
Comment (nolanlawson) [inline packages/@lwc/engine-core/src/framework/vnodes.ts:78]: Vue uses [a patch flag](https://github.com/vuejs/core/blob/245230e135152900189f13a4281302de45fdcfaa/packages/shared/src/patchFlags.ts#L114) to indicate that an element is hoisted/static. Maybe we should start using a system like this? Presumably it's more efficient than attaching members to an object.

File: packages/@lwc/engine-core/src/framework/rendering.ts [hunk 12]
```diff
@@ -172,20 +174,49 @@ function mountElement(vnode: VElement, parent: ParentNode, anchor: Node | null)
...
     vnode.elm = elm;
 
     patchElementPropsAndAttrs(null, vnode);
 
     insertNode(elm, parent, anchor);
     mountVNodes(vnode.children, elm, null);
+
+    if (isStatic) {
+        // clone the rendered node in case it is manually manipulated.
+        vnode.protoElm = cloneNode(elm, true);
```
Comment (jmsjtu) [inline packages/@lwc/engine-core/src/framework/rendering.ts:218]: I'm not sure I completely understand the usage of the `vnode.protoElm`, from what I understand it looks like we keep a reference to elm in both `vnode.protoElm` and `vnode.elm` and if we call mountElement against the `vnode` again, we replace the reference to `vnode.elm` with `vnode.protoElm`.  

I'm not super familiar with this area but could do you explain why we clone each time?

**Comment by jodarove** [comment]
closed in favor of #2781
